File: WEEK1/RC4.c
#include <stdio.h>
#include <string.h>

void swap(unsigned char *a, unsigned char *b) {
    unsigned char temp = *a;
    *a = *b;
    *b = temp;
}

void ksa(unsigned char *S, unsigned char *key, int key_len) {
    int i, j = 0;
    for (i = 0; i < 256; i++) {
        S[i] = i;
    }
    for (i = 0; i < 256; i++) {
        j = (j + S[i] + key[i % key_len]) % 256;
        swap(&S[i], &S[j]);
    }
}

void prga(unsigned char *S, unsigned char *plaintext, unsigned char *ciphertext, int len) {
    int i = 0, j = 0, k;
    for (k = 0; k < len; k++) {
        i = (i + 1) % 256;
        j = (j + S[i]) % 256;
        swap(&S[i], &S[j]);
        int t = (S[i] + S[j]) % 256;
        ciphertext[k] = S[t] ^ plaintext[k];
    }
}

void rc4_encrypt(unsigned char *plaintext, unsigned char *key, int plaintext_len, int key_len, unsigned char *ciphertext) {
    unsigned char S[256];
    ksa(S, key, key_len);
    prga(S, plaintext, ciphertext, plaintext_len);
}

void print_hex(unsigned char *data, int len) {
    for (int i = 0; i < len; i++) {
        printf("%02X ", data[i]);
    }
    printf("\n");
}
int main() {
    unsigned char plaintext[256];
    unsigned char key[256];
    
    printf("Nhap plaintext: ");
    fgets((char *)plaintext, sizeof(plaintext), stdin);
    plaintext[strcspn((char *)plaintext, "\n")] = 0;
    
    printf("Nhap key: ");
    fgets((char *)key, sizeof(key), stdin);
    key[strcspn((char *)key, "\n")] = 0;
    
    int plaintext_len = strlen((char *)plaintext);
    int key_len = strlen((char *)key);
    
    unsigned char ciphertext[plaintext_len];
    
    rc4_encrypt(plaintext, key, plaintext_len, key_len, ciphertext);
    
    printf("Plaintext: %s\n", plaintext);
    printf("Key: %s\n", key);
    printf("Ciphertext (hex): ");
    print_hex(ciphertext, plaintext_len);
    
    return 0;
}

File: WEEK1/RC4.py
def KSA(key):
    S = list(range(256))
    T = [key[i % len(key)] for i in range(256)]
    j = 0
    for i in range(256):
        j = (j + S[i] + T[i]) % 256
        S[i], S[j] = S[j], S[i]
    print(S)
    return S

def PRGA(S, data_length):
    i = 0
    j = 0
    keystream = []
    for _ in range(data_length):
        i = (i + 1) % 256
        j = (j + S[i]) % 256
        S[i], S[j] = S[j], S[i]
        key_byte = S[(S[i] + S[j]) % 256]
        keystream.append(key_byte)
    return keystream

def RC4_encrypt_decrypt(key, data):
    key = [ord(c) for c in key]
    S = KSA(key)
    keystream = PRGA(S, len(data))
    result = bytes(d ^ k for d, k in zip(data, keystream))
    return result

def input_data():
    key = input("Nhập key: ")
    plaintext = input("Nhập plaintext: ").encode()
    return key, plaintext

def output_result(ciphertext, decrypted):
    print("Ciphertext (hex):", ciphertext.hex())
    print("Decrypted text:", decrypted.decode())

if __name__ == "__main__":
    key, plaintext = input_data()
    ciphertext = RC4_encrypt_decrypt(key, plaintext)
    decrypted = RC4_encrypt_decrypt(key, ciphertext)
    output_result(ciphertext, decrypted)
